File: 407_Trapping_Rain_Water_II/trapRainWater.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
    	int trapRainWater(vector<vector<int> >& heightMap) {
            int m = heightMap.size();
            int n = heightMap[0].size();
            int res = 0;
            vector<vector<int> > left(m , vector<int>(n, 0));
            vector<vector<int> > right(m, vector<int>(n, 0));
            vector<vector<int> > up(m, vector<int>(n, 0));
            vector<vector<int> > down(m, vector<int>(n, 0));

            for(int i = 0; i < m; i++)
            {
                left[i][0] = heightMap[i][0];
                for(int j = 1; j < n; j++)
                {
                    left[i][j] = max(left[i][j - 1], heightMap[i][j]);
                }
            }

            for(int i = 0; i < m; i++)
            {
                right[i][n - 1] = heightMap[i][n - 1];
                for(int j = n - 2; j >= 0; j--)
                {
                    right[i][j] = max(right[i][j + 1], heightMap[i][j]);
                }
            }

            for(int j = 0; j < n; j++)
            {
                up[0][j] = heightMap[0][j];
                for(int i = 1; i < m; i++)
                {
                    up[i][j] = max(up[i - 1][j], heightMap[i][j]);
                }
            }

            for(int j = 0; j < n; j++)
            {
                down[m - 1][j] = heightMap[m - 1][j];
                for(int i = m - 2; i >= 0; i--)
                {
                    down[i][j] = max(down[i + 1][j], heightMap[i][j]);
                }
            }

            for(int i = 1; i < m - 1; i++)
            {
                for(int j = 1; j < n - 1; j++)
                {
                    res += (min(min(left[i][j], right[i][j]),
                               min(up[i][j], down[i][j])) - heightMap[i][j]);
                }
            }
            return res;
    	}
};

File: 407_Trapping_Rain_Water_II/trapRainWater.py
class Solution(object):
    def getMin(self, ac_point):
        minimum = ac_point.keys()[0]
        for k in ac_point.keys():
            if ac_point[k] < ac_point[minimum]:
                minimum = k

        return minimum
    
    def trapRainWater(self, heightMap):
        """
        :type heightMap: List[List[int]]
        :rtype: int
        """
        if not heightMap:
            return 0
        INF = 2 ** 31 - 1
        mm = len(heightMap)
        nn = len(heightMap[0])
        water = [[0] * nn for _ in xrange(mm)]
        ac_point = {}
        
        for i in xrange(mm):
            for j in xrange(nn):
                if i == 0 or i == mm - 1:
                    water[i][j] = heightMap[i][j]
                    ac_point[(i, j)] = heightMap[i][j]
                elif j == 0 or j == nn - 1:
                    water[i][j] = heightMap[i][j]
                    ac_point[(i, j)] = heightMap[i][j] 
                else:
                    water[i][j] = INF

        dd = {(0, 1), (0, -1), (-1, 0), (1, 0)}
        
        while ac_point:
            m = self.getMin(ac_point)
            ac_point.pop(m)
            y, x = m[0], m[1] # P
            for dx, dy in dd:
                if(0 <= x + dx < nn and 0 <= y + dy < mm):
                    tmp = water[y + dy][x + dx]
                    water[y + dy][x + dx] = max(heightMap[y + dy][x + dx], min(water[y + dy][x + dx], water[y][x]))
                    if water[y + dy][x + dx] != tmp:
                        ac_point[(y + dy, x + dx )] = water[y + dy][x + dx]

        print water
        res = 0
        for i in xrange(mm):
            for j in xrange(nn):
                res = max(res, water[i][j])

        return res





sol = Solution()
heightMap = [[1,3,3,1,3,2],[3,2,1,3,2,3],[3,3,3,2,3,1]]
print sol.trapRainWater(heightMap)

                
